### PR TITLE
Synchronizing the TextWriter in CompositeTraceWriter

### LINKED ISSUE #1533: Concurrency issue in CompositeTraceWriter
Seeing the following stack trace in the wild:

```
System.InvalidOperationException : Collection was modified; enumeration operation may not execute.
   at System.ThrowHelper.ThrowInvalidOperationException(ExceptionResource resource)
   at System.Collections.Generic.List`1.Enumerator.MoveNextRare()
   at System.Collections.Generic.List`1.Enumerator.MoveNext()
   at Microsoft.Azure.WebJobs.Host.CompositeTraceWriter.InvokeTraceWriters(TraceEvent traceEvent)
   at Microsoft.Azure.WebJobs.Host.CompositeTraceWriter.Trace(TraceEvent traceEvent)
   at Microsoft.Azure.WebJobs.Host.Loggers.TraceWriterFunctionInstanceLogger.LogFunctionCompletedAsync(FunctionCompletedMessage message,CancellationToken cancellationToken)
   at async Microsoft.Azure.WebJobs.Host.Executors.FunctionExecutor.TryExecuteAsync(IFunctionInstance functionInstance,CancellationToken cancellationToken)
   at async Microsoft.Azure.WebJobs.Host.Executors.ShutdownFunctionExecutor.TryExecuteAsync(IFunctionInstance instance,CancellationToken cancellationToken)
   at async Microsoft.Azure.WebJobs.JobHost.CallAsync(??)
   at async Microsoft.Azure.WebJobs.Script.Script
```

This is coming from the code [here](https://github.com/Azure/azure-webjobs-sdk/blob/7dd9681609613f79cb73cbc13d28c5161ca1d7c3/src/Microsoft.Azure.WebJobs.Host/CompositeTraceWriter.cs#L50).

### PR DESCRIPTION
**Pull request by brettsam**
Fixes #1533.

I'm suspecting that PR #1567 unmasked this issue. It's likely been hiding behind the other issue and once it was fixed, this one is now getting hit.

### CODE REVIEW

File: test/Microsoft.Azure.WebJobs.Host.UnitTests/CompositeTraceWriterTests.cs [hunk 1]
```diff
@@ -120,5 +126,51 @@ public void Constructor_CreatesCopyOfCollection()
             Assert.Equal(2, textWriter.ToString().Split(new string[] { Environment.NewLine }, StringSplitOptions.RemoveEmptyEntries).Length);
             Assert.Equal(0, t3.GetTraces().Count);
         }
+
+        [Fact]
+        public void TextWriter_IsSynchronized()
```
Comment (brettsam) [inline test/Microsoft.Azure.WebJobs.Host.UnitTests/CompositeTraceWriterTests.cs]: This test failed 100% of the time for me before the fix. Now passes 100% of the time.

File: test/Microsoft.Azure.WebJobs.Host.UnitTests/CompositeTraceWriterTests.cs [hunk 2]
```diff
@@ -21,6 +23,9 @@ public class CompositeTraceWriterTests
         public CompositeTraceWriterTests()
         {
             _mockTextWriter = new Mock<TextWriter>(MockBehavior.Strict);
+            _mockTextWriter.Setup(m => m.FormatProvider)
+                .Returns(new Mock<IFormatProvider>(MockBehavior.Strict).Object);
+
```
Comment (brettsam) [inline test/Microsoft.Azure.WebJobs.Host.UnitTests/CompositeTraceWriterTests.cs]: `TextWriter.Synchronized` needed this property to return something.

**Comment by brettsam** [comment]
@fabiocav / @maiqbal11 -- with Mathew on servicing, would one of you guys take a look at this? Should be quick.

File: src/Microsoft.Azure.WebJobs.Host/CompositeTraceWriter.cs [hunk 3]
```diff
@@ -25,7 +25,10 @@ public CompositeTraceWriter(IEnumerable<TraceWriter> traceWriters, TextWriter te
             var traceWritersCopy = traceWriters?.ToList() ?? new List<TraceWriter>();
             _innerTraceWriters = traceWritersCopy.AsReadOnly();
 
-            _innerTextWriter = textWriter;
+            if (textWriter != null)
+            {
+                _innerTextWriter = TextWriter.Synchronized(textWriter);
```
Comment (mathewc) [inline src/Microsoft.Azure.WebJobs.Host/CompositeTraceWriter.cs]: The other innerTraceWriters we're delegating to as well are already synchronized?

Comment (brettsam) [reply src/Microsoft.Azure.WebJobs.Host/CompositeTraceWriter.cs]: The other few places we use a TextWriter, we're either wrapping it in Synchronized already (I fixed a few a while back) or they're not being shared.

**Pull request merged**